compiler/rustc_mir_transform/src/coroutine/relocate_upvars.rs

```diff
@@ -0,0 +1,451 @@
+//! MIR rewrite pass to promote upvars into native locals in the coroutine body
+//!
+//! # Summary
+//! This pass performs the following transformations.
+//! 1. It generates a fresh batch of locals for each captured upvars.
+//!
+//! For each upvar, whether used or not, a fresh local is created with the same type.
+//!
+//! 2. It replaces the places pointing into those upvars with places pointing into those locals instead
+//!
+//! Each place that starts with access into the coroutine structure `_1` is replaced with the fresh local as
+//! the base. For instance, `(_1.4 as Some).0` is rewritten into `(_34 as Some).0` when `_34` is the fresh local
```

Comment:
What is the role of the `as Some` and `.0` in this example? I am a bit confused.

I think it'd be good to extend the example and show the entire transform, including the  "prologue", on a super simple MIR body.